### ISSUE TITLE
Original Python Loss Layer in Caffe is not working. 

### ISSUE THREAD

**Issue by w777kk** [BVLC/caffe#6772]
Hi, I am a beginner in caffe programming. I am trying to do 2 class segmentation using [fcn-32s](https://github.com/longjon/caffe/tree/future). Fcn-32s has softmaxwithloss layer at the end when it is trained. But I replaced it with softmax and my original dice loss layer.
I use softmax layer to turn the output value of score into probability before putting it to dice layer.
This is my dice layer code.

```
import random
import numpy as np
import caffe
import copy
class Dice(caffe.Layer):

    def setup(self, bottom, top):
        if len(bottom) != 2:
            raise Exception("Need two inputs to compute Dice coefficient.")
        # compute sum over all axes but the batch and channel axes
        self.sum_axes = tuple(range(1, bottom[0].data.ndim - 1))

    def reshape(self, bottom, top):
        # check input dimensions match
        if bottom[0].count != 2*bottom[1].count:
            raise Exception("Prediction must have twice the number of elements of the input.")
        # loss output is scalar
        top[0].reshape(1)

    def forward(self, bottom, top):
        label = copy.deepcopy(bottom[1].data[:,0,:,:])
        # compute prediction
        #bottom[0].data.shape=1,2,512,512
	#bottom[0].data[:,0,:,:].shape=1.512.512

	pred = copy.deepcopy(bottom[0].data[:,1,:,:])
        # area of predicted contour
        a_p = np.sum(pred, axis=self.sum_axes)
        # area of contour in label
        a_l = np.sum(label, axis=self.sum_axes)
        # area of intersection
        a_pl = np.sum(pred * label, axis=self.sum_axes)
        # dice coefficient
        self.dice_coeff = np.mean(2.*a_pl/(a_p + a_l))
        top[0].data[...] = (1.0 - self.dice_coeff)

    def backward(self, top, propagate_down, bottom):
        if propagate_down[1]:
            raise Exception("Label not diff")
        elif propagate_down[0]:
            bottom[0].diff[...] = top[0].diff[...]
        else:
            raise Exception("No > `diff")
```

I created this layer based on [Loss layer](https://github.com/yihui-he/caffe-dice-loss-layer/blob/master/perClassLoss.py).

I guess  the loss output is correct. The output value is between 0.0 ~ 1.0.
But It seems like the loss function does not work well.
I put loss_weight: 1 in train.prototxt file, but does not work, getting the same loss values in the same order every time I train the network.
Moreover, the output images are all zero images after finishing training.
And when I use softmaxwithloss layer, I can get proper output images.
This is the last layers of my network.

```
layer { type: 'Crop' name: 'score' top: 'score' bottom: 'upscore' bottom: 'data' }
layer { type: 'Softmax' name: 'prob' top: 'prob' bottom: 'score' }
layer {  type: "Python"  name: "dice"  bottom: "prob"  bottom: "label"  top: "dice"
  python_param {    module: "dice"    layer: "Dice"  }
  exclude { stage: "deploy" }
  loss_weight: 1
}
```

So any helps or hints are appreciated.
Thank you.
